helix-event/src/redraw.rs

```diff
@@ -0,0 +1,51 @@
+//! Signals that control when/if the editor redraws
+
+use std::future::Future;
+
+use parking_lot::{RwLock, RwLockReadGuard};
+use tokio::sync::Notify;
+
+/// A `Notify` instance that can be used to (asyronously) request
+/// the editor the render a new frame.
+static REDRAW_NOTIFY: Notify = Notify::const_new();
+
+/// A `RwLock` that prevents the next frame from being
+/// drawn until an exclusive (write) lock can be acquired.
+/// This allows asynchsonous tasks to acquire `non-exclusive`
+/// locks (read) to prevent the next frame from being drawn
+/// until a certain computation has finished.
+static RENDER_LOCK: RwLock<()> = RwLock::new(());
+
+pub type RenderLockGuard = RwLockReadGuard<'static, ()>;
+
+/// Requests that the editor is redrawn. The redraws are
+/// debounced (currently to 30FPS) so this can be called many times
+/// without causing a ton of frames to be rendered.
+pub fn request_redraw() {
+    REDRAW_NOTIFY.notify_one();
+}
+
+/// Returns a future that will yield once a redraw has been asynchronously
+/// requested using [`request_redraw`].
+pub fn redraw_requested() -> impl Future<Output = ()> {
+    REDRAW_NOTIFY.notified()
+}
+
+/// Wait until all locks acquired with [`lock_frame`] have
+/// have been released. This function is called before rendering
+/// and is intended to allow the frame to wait for async
+/// the next frame from being drawn until the returned guard
+/// is dropped.
```

Comment:
I think this is missing a word after async, not sure what though cause I am bit confused by the wording

Reply:
Yeah I had a brain fart there 😅